packages/jellyfish-transaction/src/script/dftx/dftx_icxorderbook.ts

```diff
@@ -0,0 +1,50 @@
+import { BufferComposer, ComposableBuffer } from '../../buffer/buffer_composer'
+import BigNumber from 'bignumber.js'
+import { Script } from '../../tx'
+import { CScript } from '../../tx_composer'
+
+// Disabling no-return-assign makes the code cleaner with the setter and getter */
+/* eslint-disable no-return-assign */
+
+export enum ICXOrderType {
+  /** type for DFI/BTC orders */
+  INTERNAL = 1,
+  /** type for BTC/DFI orders */
+  EXTERNAL = 2
+}
+
+/**
+ * ICX CreateOrder DeFi Transaction
+ */
+
+export interface ICXCreateOrder {
+  orderType: number // -------------| 1 byte unsigned
+  tokenId: number // ---------------| VarUInt{1-9 bytes}
+  ownerAddress: Script // ----------| n = VarUInt{1-9 bytes}, + n bytes
+  receivePubkey?: string // --------| 1 byte for len + len bytes
+  amountFrom: BigNumber // ---------| 8 bytes unsigned
+  orderPrice: BigNumber // ---------| 8 bytes unsigned
+  expiry: number // ----------------| 4 bytes unsigned
+}
+
+/**
+ * Composable ICXCreateOrder, C stands for Composable.
+ * Immutable by design, bi-directional fromBuffer, toBuffer deep composer.
+ */
+export class CICXCreateOrder extends ComposableBuffer<ICXCreateOrder> {
+  static OP_CODE = 0x31 // '1'
+  static OP_NAME = 'OP_DEFI_TX_ICX_CREATE_ORDER'
+
+  composers (cco: ICXCreateOrder): BufferComposer[] {
+    return [
+      ComposableBuffer.uInt8(() => cco.orderType, v => cco.orderType = v),
+      ComposableBuffer.varUInt(() => cco.tokenId, v => cco.tokenId = v),
+      ComposableBuffer.single<Script>(() => cco.ownerAddress, v => cco.ownerAddress = v, v => new CScript(v)),
+      ComposableBuffer.optionalVarUIntHex(() => cco.receivePubkey, v => cco.receivePubkey = v),
+      ComposableBuffer.satoshiAsBigNumber(() => cco.amountFrom, v => cco.amountFrom = v),
+      ComposableBuffer.satoshiAsBigNumber(() => cco.amountFrom, v => cco.amountFrom = v),
```

Comment:
is this duplicated or intentionally??

Reply:
oops my bad

Reply:
@Jouzo maybe should add a note why we are doing this. since `amountToFill` is not in  jelyfish side `ICXCreateOrder` interface, a reader might get confused. Also, `amountFrom` and `amountToFill` are two different properties that would not be the same for a partially filled order. but will be the same for `createOrder`
maybe also add this link in the comment
https://github.com/DeFiCh/ain/blob/ff53dcee23db2ffe0da9b147a0a53956f4e7ee31/src/masternodes/icxorder.h#L32